# Getting the postal codes and states of companies using their names and countries
-----
### 1. Google API Geocoding

##### 1. Transform the dataframe in to the proper format

In [331]:
import pandas as pd

xls = pd.ExcelFile('/Users/kitaeklee/Desktop/ListofAccounts__EN.xlsx')
xl_file = pd.read_excel(xls)



In [332]:
# remove the first row which is empty
xl_file.drop([0], axis=0, inplace=True)

In [333]:
#Reset the index
xl_file.reset_index(inplace=True, drop=True)

In [334]:
# Apply the first row as a Header
header = xl_file.iloc[0]
xl_file.columns = header

In [335]:
# Drop the first row after apply the first row as a header
xl_file.drop([0], axis=0, inplace=True)

In [336]:
# Reset the index again
xl_file.reset_index(drop=True, inplace=True)

In [337]:
# Create empty columns for latitudes and longitudes
xl_file['latitude'] = ''
xl_file['longitude'] = ''

In [338]:
xl_file['Full_Address'] = ''

Change the name of the countries such as Hong Kong, China and Taiwan, China to Hong Kong and Taiwan each for the accuracy of geocoding.

In [339]:
a=0
for country in xl_file['Country']:
    if country == 'Hong Kong, China':
        xl_file.loc[a, 'Country'] = 'Hong Kong'
        a+=1
        continue
    elif country == 'Taiwan, China':
        xl_file.loc[a, 'Country'] = 'Taiwan'
        a+=1
        continue
    else:
        a+=1

In [340]:
# create a new column 'address' to combine 'Company Name' and 'Country'
xl_file['address'] = xl_file['Account Name'] + ', ' + xl_file['Country']

In [341]:
# convert the data type to string
xl_file['address'] = xl_file['address'].astype(str)

In [342]:
xl_file

0     Status Account ID                                  Account Name  \
0     Active   10000200                                  (주)고려아카데미컨설팅   
1     Active    1000495               3D Industrial Imaging Co., Ltd.   
2     Active    1000496                                     3H SYSTEM   
3     Active    1000497                               3SSOFT CO.,LTD.   
4     Active    1000498                                 3ware (주)쓰리웨어   
..       ...        ...                                           ...   
972   Active    1000465                   ZENITH&COMPANY (주)제니스 앤 컴퍼니   
973   Active    1000187                               ZensM CO., Ltd.   
974   Active    1000122  Zing Semiconductor Corporation 上海新昇半导体科技有限公司   
975   Active   10000193                    ZPMC SMART SOLUTIONS GROUP   
976  Blocked   10000199                                北京博珥佐特软件科技有限公司   

0        Country Tier Class       Industry   Main Contact State Postal Code  \
0    South Korea          -            NaN            NaN   NaN         NaN   
1    South Korea          -            NaN            NaN   NaN         NaN   
2    South Korea          -            NaN            NaN   NaN         NaN   
3    South Korea          -            NaN            NaN   NaN         NaN   
4    South Korea          -            NaN            NaN   NaN         NaN   
..           ...        ...            ...            ...   ...         ...   
972  South Korea          -            NaN            NaN   NaN         NaN   
973  South Korea          -            NaN            NaN   NaN         NaN   
974        China          -  Semiconductor       JiangLan   NaN         NaN   
975        China          -            NaN  Wang, Zhenguo   NaN         NaN   
976  South Korea          -            NaN            NaN   NaN         NaN   

0   latitude longitude Full_Address  \
0                                     
1                                     
2                                     
3                                     
4                                     
..       ...       ...          ...   
972                                   
973                                   
974                                   
975                                   
976                                   

0                                              address  
0                            (주)고려아카데미컨설팅, South Korea  
1         3D Industrial Imaging Co., Ltd., South Korea  
2                               3H SYSTEM, South Korea  
3                         3SSOFT CO.,LTD., South Korea  
4                           3ware (주)쓰리웨어, South Korea  
..                                                 ...  
972           ZENITH&COMPANY (주)제니스 앤 컴퍼니, South Korea  
973                       ZensM CO., Ltd., South Korea  
974  Zing Semiconductor Corporation 上海新昇半导体科技有限公司, ...  
975                  ZPMC SMART SOLUTIONS GROUP, China  
976                        北京博珥佐特软件科技有限公司, South Korea  

[977 rows x 13 columns]

##### 2. Use Google API to get states and postal codes firstly

In [314]:
pip install requests

Note: you may need to restart the kernel to use updated packages.


Below is Def for extracting longitudes and latitudes from Google Geocoding API

In [343]:
import requests

GOOGLE_API_KEY = 'AIzaSyBC1EZSYwCfI9R4opYvLXPG3PBPBQwz_0o' 

def extract_lat_long_via_address(address_or_zipcode):
    lat, lng = None, None
    api_key = GOOGLE_API_KEY
    base_url = "https://maps.googleapis.com/maps/api/geocode/json"
    endpoint = f"{base_url}?address={address_or_zipcode}&key={api_key}"
   
    r = requests.get(endpoint)
#    if r.status_code not in range(200, 299):
#        return None, None
    
    post_code, address, state = None, None, None
    
    try:
        '''
        This try block in case any of our inputs are invalid. This is done instead
        of actually writing out handlers for all kinds of responses.
        '''
        results = r.json()['results'][0]
        
        #1
        lat = results['geometry']['location']['lat']
        #2
        lng = results['geometry']['location']['lng']
       
        #3
        for i in results['address_components']:
            if i['types'][0] == 'postal_code':
                post_code = i['long_name']   
        
        #4
        if 'formatted_address' in results:
            address = results['formatted_address']
        
        #5
        for i in results['address_components']:
            if i['types'][0] == 'administrative_area_level_1':
                state = i['long_name']

                
    except:
        pass
    return lat, lng, post_code, address, state


Example of output for one company using Google API and extract_lat_long_via_address():

In [344]:

lat, lng = None, None
api_key = GOOGLE_API_KEY
address_or_zipcode = '(주)고려아카데미컨설팅, South Korea'
base_url = "https://maps.googleapis.com/maps/api/geocode/json"
endpoint = f"{base_url}?address={address_or_zipcode}&key={api_key}"
   
r = requests.get(endpoint)

results = r.json()['results'][0]

results   



{'address_components': [{'long_name': '32',
   'short_name': '32',
   'types': ['premise']},
  {'long_name': 'Digital-ro 9-gil',
   'short_name': 'Digital-ro 9-gil',
   'types': ['political', 'sublocality', 'sublocality_level_4']},
  {'long_name': 'Gasan-dong',
   'short_name': 'Gasan-dong',
   'types': ['political', 'sublocality', 'sublocality_level_2']},
  {'long_name': 'Geumcheon-gu',
   'short_name': 'Geumcheon-gu',
   'types': ['political', 'sublocality', 'sublocality_level_1']},
  {'long_name': 'Seoul',
   'short_name': 'Seoul',
   'types': ['administrative_area_level_1', 'political']},
  {'long_name': 'South Korea',
   'short_name': 'KR',
   'types': ['country', 'political']},
  {'long_name': '153-023', 'short_name': '153-023', 'types': ['postal_code']}],
 'formatted_address': '32 Digital-ro 9-gil, Gasan-dong, Geumcheon-gu, Seoul, South Korea',
 'geometry': {'location': {'lat': 37.479449, 'lng': 126.887477},
  'location_type': 'ROOFTOP',
  'viewport': {'northeast': {'lat': 37.48

In [345]:
for country in xl_file['Country'].isnull():
    if country == True:
        print("True")

True
True
True
True
True
True
True
True
True
True
True
True
True
True


--> There are rows with no country provided from customers initially which cause Google API to return inaccurate information. Therefore, while looping through the dataframe, I will skip the rows with no country provided.

Loop through the 'address' and use Google Geocoding API to get latitude and longitude then put them in 'latitude' and 'longitude' columns.

It takes approximately 1 second to get one pair of latitude and longitude.


In [346]:
z=0
for address, country in zip(xl_file['address'], xl_file['Country'].isnull()):
    if country == False:
        dic = extract_lat_long_via_address(address)

        xl_file.loc[z, 'latitude'] = dic[0]
        xl_file.loc[z, 'longitude'] = dic[1]
        xl_file.loc[z, 'Postal Code'] = dic[2]
        xl_file.loc[z, 'Full_Address'] = dic[3]
        xl_file.loc[z, 'State'] = dic[4]


        z+=1
    else:
        z+=1


In [347]:
xl_file

0     Status Account ID                                  Account Name  \
0     Active   10000200                                  (주)고려아카데미컨설팅   
1     Active    1000495               3D Industrial Imaging Co., Ltd.   
2     Active    1000496                                     3H SYSTEM   
3     Active    1000497                               3SSOFT CO.,LTD.   
4     Active    1000498                                 3ware (주)쓰리웨어   
..       ...        ...                                           ...   
972   Active    1000465                   ZENITH&COMPANY (주)제니스 앤 컴퍼니   
973   Active    1000187                               ZensM CO., Ltd.   
974   Active    1000122  Zing Semiconductor Corporation 上海新昇半导体科技有限公司   
975   Active   10000193                    ZPMC SMART SOLUTIONS GROUP   
976  Blocked   10000199                                北京博珥佐特软件科技有限公司   

0        Country Tier Class       Industry   Main Contact     State  \
0    South Korea          -            NaN            NaN     Seoul   
1    South Korea          -            NaN            NaN      None   
2    South Korea          -            NaN            NaN      None   
3    South Korea          -            NaN            NaN     Seoul   
4    South Korea          -            NaN            NaN     Seoul   
..           ...        ...            ...            ...       ...   
972  South Korea          -            NaN            NaN  Missouri   
973  South Korea          -            NaN            NaN      None   
974        China          -  Semiconductor       JiangLan  Shanghai   
975        China          -            NaN  Wang, Zhenguo      None   
976  South Korea          -            NaN            NaN   Beijing   

0   Postal Code latitude longitude  \
0       153-023  37.4794   126.887   
1          None     None      None   
2          None     None      None   
3       137-130  37.4754   127.042   
4       153-023  37.4813   126.884   
..          ...      ...       ...   
972       65804  37.1779   -93.222   
973        None     None      None   
974        None  31.2304   121.474   
975        None     None      None   
976        None  39.9042   116.407   

0                                         Full_Address  \
0    32 Digital-ro 9-gil, Gasan-dong, Geumcheon-gu,...   
1                                                 None   
2                                                 None   
3      262 Yangjae-dong, Seocho-gu, Seoul, South Korea   
4    50-3 Gasan-dong, Geumcheon-gu, Seoul, South Korea   
..                                                 ...   
972  3534 E Sunshine St #3, Springfield, MO 65804, USA   
973                                               None   
974                                    Shanghai, China   
975                                               None   
976                                     Beijing, China   

0                                              address  
0                            (주)고려아카데미컨설팅, South Korea  
1         3D Industrial Imaging Co., Ltd., South Korea  
2                               3H SYSTEM, South Korea  
3                         3SSOFT CO.,LTD., South Korea  
4                           3ware (주)쓰리웨어, South Korea  
..                                                 ...  
972           ZENITH&COMPANY (주)제니스 앤 컴퍼니, South Korea  
973                       ZensM CO., Ltd., South Korea  
974  Zing Semiconductor Corporation 上海新昇半导体科技有限公司, ...  
975                  ZPMC SMART SOLUTIONS GROUP, China  
976                        北京博珥佐特软件科技有限公司, South Korea  

[977 rows x 13 columns]

In [348]:
xl_file['latitude'].notnull().sum()

752

--> 752 out of 977 companies got latitudes returned.

In [349]:
xl_file['State'].notnull().sum()

704

--> 704 out of 977 companies got state returned.

In [350]:
xl_file['Postal Code'].notnull().sum()

556

--> 556 out of 977 companies got postal codes returned.

----
##### Data Cleaning

There are rows with wrong information returned. Therefore, I will check if intially given customers' countries are same as countries returned from Google API. If not, I will consider that as inaccurate information and remove all the information returned from Google API for those corresponding information of companies.

To check that information, I will check if initially given country is contained in the returned full address of each company. The name of countries to check are not only its full name but also abbreviated version. For example, 'Argentina' can also be 'AR' and 'ARG' therefore if any one of those three are included in the address of a company in Argentina, it will be considered as accurate information.

In [473]:
xl_file2 = xl_file.copy()

There are customers who did not provide their country. Their empty cells of latitude and longitude cause error later therefore I will apply 'None' to them.

In [458]:
xl_file2[xl_file2['Country'].isnull()]

0     Status Account ID                     Account Name Country Tier Class  \
43    Active    1000688                   AKT Components     NaN          -   
165   Active    1000699           CollabraTech Solutions     NaN          -   
389  Blocked    1000721                      INFICON/FPS     NaN          -   
531  Blocked    1000730       Mega Kinetic Fluid Systems     NaN          -   
541  Blocked    1000733                             MICL     NaN          -   
594  Blocked    1000739                              NXP     NaN          -   
660  Blocked    1000748                            Qorvo     NaN          -   
671   Active    1000749                          Renesas     NaN          -   
735   Active    1000755  SEMI/Group Purchasing Solutions     NaN          -   
736   Active    1000756   Semiconductor Global Solutions     NaN          -   
803  Blocked    1000762                         Skyworks     NaN          -   
845  Blocked    1000767                           Tanaka     NaN          -   
854  Blocked    1000769                              TEL     NaN          -   
883  Blocked    1000772                        TowerJazz     NaN          -   

0   Industry     Main Contact State Postal Code latitude longitude  \
43       NaN     Douglas, Dan   NaN         NaN                      
165      NaN       Ptak, John   NaN         NaN                      
389      NaN              NaN   NaN         NaN                      
531      NaN              NaN   NaN         NaN                      
541      NaN     Shyken, Koby   NaN         NaN                      
594      NaN              NaN   NaN         NaN                      
660      NaN              NaN   NaN         NaN                      
671      NaN              NaN   NaN         NaN                      
735      NaN              NaN   NaN         NaN                      
736      NaN     Park, Andrew   NaN         NaN                      
803      NaN              NaN   NaN         NaN                      
845      NaN  Ihara, Yasutaka   NaN         NaN                      
854      NaN              NaN   NaN         NaN                      
883      NaN              NaN   NaN         NaN                      

0   Full_Address address  
43                   nan  
165                  nan  
389                  nan  
531                  nan  
541                  nan  
594                  nan  
660                  nan  
671                  nan  
735                  nan  
736                  nan  
803                  nan  
845                  nan  
854                  nan  
883                  nan

--> These are the customers who did not provide country. Create a list of indexes of them.

In [475]:
country_null = [43, 165, 389, 531, 541, 594, 660, 671, 735, 736, 803, 845, 854, 883]

for idx in country_null:
    xl_file2.loc[idx, 'latitude'] = None
    xl_file2.loc[idx, 'longitude'] = None
    xl_file2.loc[idx, 'Full_Address'] = None
    xl_file2.loc[idx, 'State'] = None
    xl_file2.loc[idx, 'Postal Code'] = None
    xl_file2.loc[idx, 'Country'] = None
    xl_file2.loc[idx, 'address'] = None


In [476]:
xl_file2[xl_file2['Country'].isnull()]

0     Status Account ID                     Account Name Country Tier Class  \
43    Active    1000688                   AKT Components    None          -   
165   Active    1000699           CollabraTech Solutions    None          -   
389  Blocked    1000721                      INFICON/FPS    None          -   
531  Blocked    1000730       Mega Kinetic Fluid Systems    None          -   
541  Blocked    1000733                             MICL    None          -   
594  Blocked    1000739                              NXP    None          -   
660  Blocked    1000748                            Qorvo    None          -   
671   Active    1000749                          Renesas    None          -   
735   Active    1000755  SEMI/Group Purchasing Solutions    None          -   
736   Active    1000756   Semiconductor Global Solutions    None          -   
803  Blocked    1000762                         Skyworks    None          -   
845  Blocked    1000767                           Tanaka    None          -   
854  Blocked    1000769                              TEL    None          -   
883  Blocked    1000772                        TowerJazz    None          -   

0   Industry     Main Contact State Postal Code latitude longitude  \
43       NaN     Douglas, Dan  None        None     None      None   
165      NaN       Ptak, John  None        None     None      None   
389      NaN              NaN  None        None     None      None   
531      NaN              NaN  None        None     None      None   
541      NaN     Shyken, Koby  None        None     None      None   
594      NaN              NaN  None        None     None      None   
660      NaN              NaN  None        None     None      None   
671      NaN              NaN  None        None     None      None   
735      NaN              NaN  None        None     None      None   
736      NaN     Park, Andrew  None        None     None      None   
803      NaN              NaN  None        None     None      None   
845      NaN  Ihara, Yasutaka  None        None     None      None   
854      NaN              NaN  None        None     None      None   
883      NaN              NaN  None        None     None      None   

0   Full_Address address  
43          None    None  
165         None    None  
389         None    None  
531         None    None  
541         None    None  
594         None    None  
660         None    None  
671         None    None  
735         None    None  
736         None    None  
803         None    None  
845         None    None  
854         None    None  
883         None    None

--> None are applied to these rows with corresponding columns.

In [477]:
t=0
for address, country in zip(xl_file2['Full_Address'], xl_file2['Country']):
    if country is not None:
        if address is not None:
            if country == 'Argentina':
                if 'AR' in address:
                    t+=1
                elif 'ARG' in address:
                    t+=1
                elif 'Argentina' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Australia':
                if 'AU' in address:
                    t+=1
                elif 'AUS' in address:
                    t+=1
                elif 'Australia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Austria':
                if 'AT' in address:
                    t+=1
                elif 'AUT' in address:
                    t+=1
                elif 'Austria' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Bahrain':
                if 'BH' in address:
                    t+=1
                elif 'BHR' in address:
                    t+=1
                elif 'Bahrain' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Belgium':
                if 'BE' in address:
                    t+=1
                elif 'BEL' in address:
                    t+=1
                elif 'Belgium' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Bhutan':
                if 'BT' in address:
                    t+=1
                elif 'BTN' in address:
                    t+=1
                elif 'Bhutan' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Botswana':
                if 'BW' in address:
                    t+=1
                elif 'BWA' in address:
                    t+=1
                elif 'Botswa' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Brazil':
                if 'BR' in address:
                    t+=1
                elif 'BRA' in address:
                    t+=1
                elif 'Brazil' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Cameroon':
                if 'CM' in address:
                    t+=1
                elif 'CMR' in address:
                    t+=1
                elif 'Cameroon' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Canada':
                if 'CA' in address:
                    t+=1
                elif 'CAN' in address:
                    t+=1
                elif 'Canada' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Chile':
                if 'CL' in address:
                    t+=1
                elif 'CHL' in address:
                    t+=1
                elif 'Chile' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'China':
                if 'CN' in address:
                    t+=1
                elif 'CHN' in address:
                    t+=1
                elif 'China' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Colombia':
                if 'CO' in address:
                    t+=1
                elif 'COL' in address:
                    t+=1
                elif 'Colombia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Denmark':
                if 'DK' in address:
                    t+=1
                elif 'DNK' in address:
                    t+=1
                elif 'Denmark' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Ecuador':
                if 'EC' in address:
                    t+=1
                elif 'ECU' in address:
                    t+=1
                elif 'Ecuador' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Egypt':
                if 'EG' in address:
                    t+=1
                elif 'EGY' in address:
                    t+=1
                elif 'Egypt' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Finland':
                if 'FI' in address:
                    t+=1
                elif 'FIN' in address:
                    t+=1
                elif 'Finland' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'France':
                if 'FR' in address:
                    t+=1
                elif 'FRA' in address:
                    t+=1
                elif 'France' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Germany':
                if 'DE' in address:
                    t+=1
                elif 'DEU' in address:
                    t+=1
                elif 'Germany' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Greece':
                if 'GR' in address:
                    t+=1
                elif 'GRC' in address:
                    t+=1
                elif 'Greece' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Hong Kong':
                if 'HK' in address:
                    t+=1
                elif 'Hong Kong' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'India':
                if 'IN' in address:
                    t+=1
                elif 'IND' in address:
                    t+=1
                elif 'India' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1 

            elif country == 'Indonesia':
                if 'ID' in address:
                    t+=1
                elif 'IDN' in address:
                    t+=1
                elif 'Indonesia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Ireland':
                if 'IE' in address:
                    t+=1
                elif 'IRL' in address:
                    t+=1
                elif 'Ireland' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Israel':
                if 'IL' in address:
                    t+=1
                elif 'ISR' in address:
                    t+=1
                elif 'Israel' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Italy':
                if 'IT' in address:
                    t+=1
                elif 'ITA' in address:
                    t+=1
                elif 'Italy' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Japan':
                if 'JP' in address:
                    t+=1
                elif 'JPN' in address:
                    t+=1
                elif 'Japan' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Kuwait':
                if 'KW' in address:
                    t+=1
                elif 'KWT' in address:
                    t+=1
                elif 'Kuwait' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Lithuania':
                if 'LT' in address:
                    t+=1
                elif 'LTU' in address:
                    t+=1
                elif 'Lithuania' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Malaysia':
                if 'MY' in address:
                    t+=1
                elif 'MYS' in address:
                    t+=1
                elif 'Malaysia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Mexico':
                if 'MX' in address:
                    t+=1
                elif 'MEX' in address:
                    t+=1
                elif 'Mexico' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Montenegro':
                if 'ME' in address:
                    t+=1
                elif 'MNE' in address:
                    t+=1
                elif 'Montenegro' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Morocco':
                if 'MA' in address:
                    t+=1
                elif 'MAR' in address:
                    t+=1
                elif 'Morocco' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Netherlands':
                if 'NL' in address:
                    t+=1
                elif 'NLD' in address:
                    t+=1
                elif 'Netherlands' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1


            elif country == 'Nigeria':
                if 'NG' in address:
                    t+=1
                elif 'NGA' in address:
                    t+=1
                elif 'Nigeria' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Norway':
                if 'NO' in address:
                    t+=1
                elif 'NOR' in address:
                    t+=1
                elif 'Norway' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1 

            elif country == 'Oman':
                if 'OM' in address:
                    t+=1
                elif 'OMN' in address:
                    t+=1
                elif 'Oman' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Pakistan':
                if 'PK' in address:
                    t+=1
                elif 'PAK' in address:
                    t+=1
                elif 'Pakistan' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Peru':
                if 'PE' in address:
                    t+=1
                elif 'PER' in address:
                    t+=1
                elif 'Peru' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Philippines':
                if 'PH' in address:
                    t+=1
                elif 'PHL' in address:
                    t+=1
                elif 'Philippines' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Poland':
                if 'PL' in address:
                    t+=1
                elif 'POL' in address:
                    t+=1
                elif 'Poland' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Portugal':
                if 'PT' in address:
                    t+=1
                elif 'PRT' in address:
                    t+=1
                elif 'Portugal' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Puerto Rico':
                if 'PR' in address:
                    t+=1
                elif 'PRI' in address:
                    t+=1
                elif 'Puerto Rico' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Qatar':
                if 'QA' in address:
                    t+=1
                elif 'QAT' in address:
                    t+=1
                elif 'Qatar' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Russian Federation':
                if 'RU' in address:
                    t+=1
                elif 'RUS' in address:
                    t+=1
                elif 'Russian Federation' in address:
                    t+=1
                elif 'Russia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Saudi Arabia':
                if 'SA' in address:
                    t+=1
                elif 'SAU' in address:
                    t+=1
                elif 'Saudi Arabia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Singapore':
                if 'SG' in address:
                    t+=1
                elif 'SGP' in address:
                    t+=1
                elif 'Singapore' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'South Africa':
                if 'ZA' in address:
                    t+=1
                elif 'ZAF' in address:
                    t+=1
                elif 'South Africa' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'South Korea':
                if 'KR' in address:
                    t+=1
                elif 'KOR' in address:
                    t+=1
                elif 'Korea' in address:
                    t+=1
                elif 'South Korea' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Spain':
                if 'ES' in address:
                    t+=1
                elif 'ESP' in address:
                    t+=1
                elif 'Spain' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Sri Lanka':
                if 'LK' in address:
                    t+=1
                elif 'LKA' in address:
                    t+=1
                elif 'Sri Lanka' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Sweden':
                if 'SE' in address:
                    t+=1
                elif 'SWE' in address:
                    t+=1
                elif 'Sweden' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Switzerland':
                if 'CH' in address:
                    t+=1
                elif 'CHE' in address:
                    t+=1
                elif 'Switzerland' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Taiwan':
                if 'TW' in address:
                    t+=1
                elif 'Taiwan' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Thailand':
                if 'TH' in address:
                    t+=1
                elif 'THA' in address:
                    t+=1
                elif 'Thailand' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Tunisia':
                if 'TN' in address:
                    t+=1
                elif 'TUN' in address:
                    t+=1
                elif 'Tunisia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Ukraine':
                if 'UA' in address:
                    t+=1
                elif 'UKR' in address:
                    t+=1
                elif 'Ukraine' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'United Arab Emirates':
                if 'AE' in address:
                    t+=1
                elif 'ARE' in address:
                    t+=1
                elif 'United Arab Emirates' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'United Kingdom':
                if 'GB' in address:
                    t+=1
                elif 'GBR' in address:
                    t+=1
                elif 'United Kingdom' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1
                    
                    
            elif country == 'United States':
                if 'US' in address:
                    t+=1
                elif 'USA' in address:
                    t+=1
                elif 'United States' in address:
                    t+=1
                elif 'United States of America' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Venezuela':
                if 'VE' in address:
                    t+=1
                elif 'VEN' in address:
                    t+=1
                elif 'Venezuela' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Vietnam':
                if 'VN' in address:
                    t+=1
                elif 'VNM' in address:
                    t+=1
                elif 'Vietnam' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1

            elif country == 'Zambia':
                if 'ZM' in address:
                    t+=1
                elif 'ZMB' in address:
                    t+=1
                elif 'Zambia' in address:
                    t+=1
                else:
                    xl_file2.loc[t, 'State'] = None
                    xl_file2.loc[t, 'Postal Code'] = None
                    xl_file2.loc[t, 'latitude'] = None
                    xl_file2.loc[t, 'longitude'] = None
                    xl_file2.loc[t, 'Full_Address'] = None
                    t+=1
            else:
                t+=1
        else:
            t+=1
    else:
        t+=1

In [478]:
xl_file2

0     Status Account ID                                  Account Name  \
0     Active   10000200                                  (주)고려아카데미컨설팅   
1     Active    1000495               3D Industrial Imaging Co., Ltd.   
2     Active    1000496                                     3H SYSTEM   
3     Active    1000497                               3SSOFT CO.,LTD.   
4     Active    1000498                                 3ware (주)쓰리웨어   
..       ...        ...                                           ...   
972   Active    1000465                   ZENITH&COMPANY (주)제니스 앤 컴퍼니   
973   Active    1000187                               ZensM CO., Ltd.   
974   Active    1000122  Zing Semiconductor Corporation 上海新昇半导体科技有限公司   
975   Active   10000193                    ZPMC SMART SOLUTIONS GROUP   
976  Blocked   10000199                                北京博珥佐特软件科技有限公司   

0        Country Tier Class       Industry   Main Contact     State  \
0    South Korea          -            NaN            NaN     Seoul   
1    South Korea          -            NaN            NaN      None   
2    South Korea          -            NaN            NaN      None   
3    South Korea          -            NaN            NaN     Seoul   
4    South Korea          -            NaN            NaN     Seoul   
..           ...        ...            ...            ...       ...   
972  South Korea          -            NaN            NaN      None   
973  South Korea          -            NaN            NaN      None   
974        China          -  Semiconductor       JiangLan  Shanghai   
975        China          -            NaN  Wang, Zhenguo      None   
976  South Korea          -            NaN            NaN      None   

0   Postal Code latitude longitude  \
0       153-023  37.4794   126.887   
1          None     None      None   
2          None     None      None   
3       137-130  37.4754   127.042   
4       153-023  37.4813   126.884   
..          ...      ...       ...   
972        None     None      None   
973        None     None      None   
974        None  31.2304   121.474   
975        None     None      None   
976        None     None      None   

0                                         Full_Address  \
0    32 Digital-ro 9-gil, Gasan-dong, Geumcheon-gu,...   
1                                                 None   
2                                                 None   
3      262 Yangjae-dong, Seocho-gu, Seoul, South Korea   
4    50-3 Gasan-dong, Geumcheon-gu, Seoul, South Korea   
..                                                 ...   
972                                               None   
973                                               None   
974                                    Shanghai, China   
975                                               None   
976                                               None   

0                                              address  
0                            (주)고려아카데미컨설팅, South Korea  
1         3D Industrial Imaging Co., Ltd., South Korea  
2                               3H SYSTEM, South Korea  
3                         3SSOFT CO.,LTD., South Korea  
4                           3ware (주)쓰리웨어, South Korea  
..                                                 ...  
972           ZENITH&COMPANY (주)제니스 앤 컴퍼니, South Korea  
973                       ZensM CO., Ltd., South Korea  
974  Zing Semiconductor Corporation 上海新昇半导体科技有限公司, ...  
975                  ZPMC SMART SOLUTIONS GROUP, China  
976                        北京博珥佐特软件科技有限公司, South Korea  

[977 rows x 13 columns]

In [479]:
xl_file2['latitude'].notnull().sum()

635

--> Information of 117 companies are removed from the first data frame due to inaccuracy in addresses returned from Google Geocoding API.

In [480]:
xl_file2['State'].notnull().sum()

602

--> Information of 102 companies are removed from the first data frame due to inaccuracy in addresses returned from Google Geocoding API.

In [481]:
xl_file2['Postal Code'].notnull().sum()

463

--> Information of 93 companies are removed from the first data frame due to inaccuracy in addresses returned from Google Geocoding API.

---
### 2. Reversing Latitudes and Longitudes using Mapbox API

Since there are companies with latitudes and longitudes returned but not State or Postal Code, this time I will reverse latitudes and longitudes using Mapbox API to see if there are any more firms I can get State or Postal Code returned from geocoding API.



In [510]:
xl_file3 = xl_file2.copy()

In [511]:
import geocoder

#key = token from Mapbox
z=0
for lat, lng, postal, state in zip(xl_file3['latitude'], xl_file3['longitude'], xl_file3['Postal Code'], xl_file3['State']):  
    if lat is not None:
        latlng = [lat, lng]
        g = geocoder.mapbox(latlng, method='reverse', key='pk.eyJ1IjoieG9yOTIzNyIsImEiOiJja2hsYTRydHAxa3l4MnJucXB0YXZzeXFyIn0._r9Zw8FZ4rvarTcrcCJf-w')
                
        if postal is None:
            if 'postal' in g.json:
                xl_file3.loc[z, 'Postal Code'] = g.json['postal']
                z+=1
        elif state is None:
            if 'state' in g.json:
                xl_file3.loc[z,'State'] = g.json['state']
                z+=1
        else:
            z+=1
    else:
        z+=1
        



In [512]:
xl_file3['State'].notnull().sum()


616

--> Mapbox API returend states of 14 more companies initially with latitude but not the state.

In [513]:
xl_file3['Postal Code'].notnull().sum()


472

--> Mapbox API returned postal codes of 9 more companies intially with latitude but not the postal code.

In [514]:
xl_file3

0     Status Account ID                                  Account Name  \
0     Active   10000200                                  (주)고려아카데미컨설팅   
1     Active    1000495               3D Industrial Imaging Co., Ltd.   
2     Active    1000496                                     3H SYSTEM   
3     Active    1000497                               3SSOFT CO.,LTD.   
4     Active    1000498                                 3ware (주)쓰리웨어   
..       ...        ...                                           ...   
972   Active    1000465                   ZENITH&COMPANY (주)제니스 앤 컴퍼니   
973   Active    1000187                               ZensM CO., Ltd.   
974   Active    1000122  Zing Semiconductor Corporation 上海新昇半导体科技有限公司   
975   Active   10000193                    ZPMC SMART SOLUTIONS GROUP   
976  Blocked   10000199                                北京博珥佐特软件科技有限公司   

0        Country Tier Class       Industry   Main Contact     State  \
0    South Korea          -            NaN            NaN     Seoul   
1    South Korea          -            NaN            NaN      None   
2    South Korea          -            NaN            NaN      None   
3    South Korea          -            NaN            NaN     Seoul   
4    South Korea          -            NaN            NaN     Seoul   
..           ...        ...            ...            ...       ...   
972  South Korea          -            NaN            NaN      None   
973  South Korea          -            NaN            NaN      None   
974        China          -  Semiconductor       JiangLan  Shanghai   
975        China          -            NaN  Wang, Zhenguo      None   
976  South Korea          -            NaN            NaN      None   

0   Postal Code latitude longitude  \
0       153-023  37.4794   126.887   
1          None     None      None   
2          None     None      None   
3       137-130  37.4754   127.042   
4       153-023  37.4813   126.884   
..          ...      ...       ...   
972        None     None      None   
973        None     None      None   
974        None  31.2304   121.474   
975        None     None      None   
976        None     None      None   

0                                         Full_Address  \
0    32 Digital-ro 9-gil, Gasan-dong, Geumcheon-gu,...   
1                                                 None   
2                                                 None   
3      262 Yangjae-dong, Seocho-gu, Seoul, South Korea   
4    50-3 Gasan-dong, Geumcheon-gu, Seoul, South Korea   
..                                                 ...   
972                                               None   
973                                               None   
974                                    Shanghai, China   
975                                               None   
976                                               None   

0                                              address  
0                            (주)고려아카데미컨설팅, South Korea  
1         3D Industrial Imaging Co., Ltd., South Korea  
2                               3H SYSTEM, South Korea  
3                         3SSOFT CO.,LTD., South Korea  
4                           3ware (주)쓰리웨어, South Korea  
..                                                 ...  
972           ZENITH&COMPANY (주)제니스 앤 컴퍼니, South Korea  
973                       ZensM CO., Ltd., South Korea  
974  Zing Semiconductor Corporation 上海新昇半导体科技有限公司, ...  
975                  ZPMC SMART SOLUTIONS GROUP, China  
976                        北京博珥佐特软件科技有限公司, South Korea  

[977 rows x 13 columns]

---
### 3. Mapbox API Geocoding
Loop through the 'address' one more time using Mapbox API for rows with no information or inaccurate information returned from Google API to see if I can get any more address information for those companies.

In [515]:
xl_file4 = xl_file3.copy()

In [165]:
g_sample = geocoder.mapbox(xl_file4.loc[50, 'address'], key='pk.eyJ1IjoieG9yOTIzNyIsImEiOiJja2hsYTRydHAxa3l4MnJucXB0YXZzeXFyIn0._r9Zw8FZ4rvarTcrcCJf-w')

g_sample.json['country'].strip()


'United States'

In [516]:
import geocoder
c=0
for latitude, country, full_add in zip(xl_file4['latitude'], xl_file4['Country'], xl_file4['Full_Address']):
    try:
        if lat is None:
            g1 = geocoder.mapbox(xl_file4.loc[c, 'address'], key='pk.eyJ1IjoieG9yOTIzNyIsImEiOiJja2hsYTRydHAxa3l4MnJucXB0YXZzeXFyIn0._r9Zw8FZ4rvarTcrcCJf-w')
        
            if type(g1.json) is None:
                c+=1
                
            elif country != g1.json['country'].strip():
                c+=1
        
            else:
                if 'lat' in g1.json:
                    xl_file4.loc[c, 'latitude'] = g1.json['lat']
                else:
                    pass
            
                if 'lng' in g1.json:
                    xl_file4.loc[c, 'longitude'] = g1.json['lng']
                else:
                    pass
        
                if 'state' in g1.json:
                    xl_file4.loc[c, 'State'] = g1.json['state']
                else:
                    pass
        
                if 'postal' in g1.json:
                    xl_file4.loc[c, 'Postal Code'] = g1.json['postal']
                else:
                    pass
                c+=1        
        else:
            c+=1
            
    except:
        pass
    


In [517]:
xl_file4['latitude'].notnull().sum()

635

In [518]:
xl_file4['State'].notnull().sum()

616

In [519]:
xl_file4['Postal Code'].notnull().sum()

472

In this case, Mapbox API did not return any additional addresses for remaining companies with no address returned from Google API.

In [520]:
xl_file4

0     Status Account ID                                  Account Name  \
0     Active   10000200                                  (주)고려아카데미컨설팅   
1     Active    1000495               3D Industrial Imaging Co., Ltd.   
2     Active    1000496                                     3H SYSTEM   
3     Active    1000497                               3SSOFT CO.,LTD.   
4     Active    1000498                                 3ware (주)쓰리웨어   
..       ...        ...                                           ...   
972   Active    1000465                   ZENITH&COMPANY (주)제니스 앤 컴퍼니   
973   Active    1000187                               ZensM CO., Ltd.   
974   Active    1000122  Zing Semiconductor Corporation 上海新昇半导体科技有限公司   
975   Active   10000193                    ZPMC SMART SOLUTIONS GROUP   
976  Blocked   10000199                                北京博珥佐特软件科技有限公司   

0        Country Tier Class       Industry   Main Contact     State  \
0    South Korea          -            NaN            NaN     Seoul   
1    South Korea          -            NaN            NaN      None   
2    South Korea          -            NaN            NaN      None   
3    South Korea          -            NaN            NaN     Seoul   
4    South Korea          -            NaN            NaN     Seoul   
..           ...        ...            ...            ...       ...   
972  South Korea          -            NaN            NaN      None   
973  South Korea          -            NaN            NaN      None   
974        China          -  Semiconductor       JiangLan  Shanghai   
975        China          -            NaN  Wang, Zhenguo      None   
976  South Korea          -            NaN            NaN      None   

0   Postal Code latitude longitude  \
0       153-023  37.4794   126.887   
1          None     None      None   
2          None     None      None   
3       137-130  37.4754   127.042   
4       153-023  37.4813   126.884   
..          ...      ...       ...   
972        None     None      None   
973        None     None      None   
974        None  31.2304   121.474   
975        None     None      None   
976        None     None      None   

0                                         Full_Address  \
0    32 Digital-ro 9-gil, Gasan-dong, Geumcheon-gu,...   
1                                                 None   
2                                                 None   
3      262 Yangjae-dong, Seocho-gu, Seoul, South Korea   
4    50-3 Gasan-dong, Geumcheon-gu, Seoul, South Korea   
..                                                 ...   
972                                               None   
973                                               None   
974                                    Shanghai, China   
975                                               None   
976                                               None   

0                                              address  
0                            (주)고려아카데미컨설팅, South Korea  
1         3D Industrial Imaging Co., Ltd., South Korea  
2                               3H SYSTEM, South Korea  
3                         3SSOFT CO.,LTD., South Korea  
4                           3ware (주)쓰리웨어, South Korea  
..                                                 ...  
972           ZENITH&COMPANY (주)제니스 앤 컴퍼니, South Korea  
973                       ZensM CO., Ltd., South Korea  
974  Zing Semiconductor Corporation 上海新昇半导体科技有限公司, ...  
975                  ZPMC SMART SOLUTIONS GROUP, China  
976                        北京博珥佐特软件科技有限公司, South Korea  

[977 rows x 13 columns]

Export the final dataframe.

In [521]:
xl_file4.to_excel(r'/Users/kitaeklee/Desktop/ListofAccounts__EN_Final.xlsx', index = False)


*** Consider looking up the address on the internet for the rows with missing postal code or state and if the returned address does not seem accurate since it might means returned address is inaccurate.***